pkg/stanza/fileconsumer/file.go

```diff
@@ -109,162 +130,161 @@ func (m *Manager) startPoller(ctx context.Context) {
 	for i := 0; i < len(m.knownFiles); i++ {
 		m.knownFiles[i].generation++
 	}
-
-	// Used to keep track of the number of batches processed in this poll cycle
-	batchesProcessed := 0
+	m.knownFilesLock.Unlock()
 
 	// Get the list of paths on disk
 	matches := m.finder.FindFiles()
-	for len(matches) > m.maxBatchFiles {
-		m.consume(ctx, matches[:m.maxBatchFiles])
+	m.consume(ctx, matches)
+	m.clearCurrentFiles()
 
-		// If a maxBatches is set, check if we have hit the limit
-		if m.maxBatches != 0 {
-			batchesProcessed++
-			if batchesProcessed >= m.maxBatches {
-				return
+	// Any new files that appear should be consumed entirely
+	m.readerFactory.fromBeginning = true
+	m.syncLastPollFiles(ctx)
+}
+
+func (m *Manager) clearCurrentFiles() {
+	m.currentFiles = make([]*os.File, 0)
+	m.currentFps = make([]*Fingerprint, 0)
+}
+
+func (m *Manager) worker(ctx context.Context) {
+	defer m.workerWg.Done()
+	for {
+		chanData, ok := <-m.readerChan
+
+		if !ok {
+			return
+		}
+		r, path := chanData.reader, chanData.path
+		r.ReadToEnd(ctx)
+		// Delete a file if deleteAfterRead is enabled and we reached the end of the file
+		if m.deleteAfterRead && r.eof {
+			r.Close()
+			if err := os.Remove(r.file.Name()); err != nil {
+				m.Errorf("could not delete %s", r.file.Name())
 			}
+		} else {
+			// Save off any files that were not fully read or if deleteAfterRead is false
+			m.saveCurrent([]*Reader{r})
+			m.readerLock.Lock()
+			m.lostReaders = append(m.lostReaders, r)
+			m.readerLock.Unlock()
 		}
-
-		matches = matches[m.maxBatchFiles:]
+		m.pathHashLock.Lock()
+		delete(m.pathHash, path)
+		m.pathHashLock.Unlock()
 	}
-	m.consume(ctx, matches)
 }
 
 func (m *Manager) consume(ctx context.Context, paths []string) {
 	m.Debug("Consuming files")
-	readers := m.makeReaders(paths)
-
-	// take care of files which disappeared from the pattern since the last poll cycle
-	// this can mean either files which were removed, or rotated into a name not matching the pattern
-	// we do this before reading existing files to ensure we emit older log lines before newer ones
-	m.roller.readLostFiles(ctx, readers)
-
-	var wg sync.WaitGroup
-	for _, reader := range readers {
-		wg.Add(1)
-		go func(r *Reader) {
-			defer wg.Done()
-			r.ReadToEnd(ctx)
-			// Delete a file if deleteAfterRead is enabled and we reached the end of the file
-			if m.deleteAfterRead && r.eof {
-				r.Close()
-				if err := os.Remove(r.file.Name()); err != nil {
-					m.Errorf("could not delete %s", r.file.Name())
-				}
-			}
-		}(reader)
+	m.handleLostFiles(ctx)
+	for _, path := range paths {
+		m.pathHashLock.Lock()
+		if _, ok := m.pathHash[path]; ok {
+			m.pathHashLock.Unlock()
+			continue
+		}
+		m.pathHash[path] = true
+		reader := m.makeReader(path)
+		if reader == nil {
+			fmt.Println("Couldn't create reader for ", path)
+			delete(m.pathHash, path)
+			m.pathHashLock.Unlock()
+			continue
+		}
+		m.pathHashLock.Unlock()
+		m.readerChan <- ReaderWrapper{reader: reader, path: path}
```

Comment:
In my opinion, this is the most critical part of the proposed design change, but the design doc does not mention it.

Documented or not, this package was designed around the idea that a file path is _not_ enough information to identify a file, so what you're proposing here requires a strong rationale that explains how it ensures no logs are missed or duplicated due to file rotation.

Reply:
Yeah, I agree regarding the rotation one. But using the `path` also ensures that we'll consume the rotated files are consumed sequentially. I'll work on finding another way.
However, there's a small amount of risk that we'll miss the logs if the rotation is occurring rapidly. 

Also, one more thing.
While performing unit testing for the rotating files, this performed as expected. All the logs were detected without being missed or duplicated.

Reply:
So I've thought of an approach:
We use a combination of the file path and fingerprint to detect the consuming files.

New Hash: map[string]fingerprint

Two scenarios:

- If the new file path doesn't exist in the hash
     - Start consuming file
-  If the new file path exists in the hash
     - Fetch the fingerprint for that path and check if it matches.
     - If the fingerprint matches, then it means it's the same file as before and we can skip it.
     - If the fingerprint doesn't match, it means it's a new file with the same name (the file is rotated) and we can start consuming it.

I'll implement it today and test it out
Let me know what do think about this!

Reply:
So I've thought of an approach: 
We use a combination of the file path and fingerprint to detect the consuming files.
New Hash: map[string]fingerprint
Two scenarios:
* If the new file path doesn't exist in the hash
    * Start consuming file.
        * **Or we can loop through the dictionary to check the fingerprints?? But this increases the time complexity and I don't think this is fruitful.**
* If the new file path exists in the hash
    * Fetch the fingerprint for that path and check if it matches.
    * If the fingerprint matches, then it means it's the same file as before and we can skip it.
    * If the fingerprint doesn't match, it means it's a new file with the same name (the file is rotated) and we can start consuming it.

I'll implement it today and test it out Let me know what do think about this!

Reply:
> So I've thought of an approach:  We use a combination of the file path and fingerprint to detect the consuming files. New Hash: map[string]fingerprint Two scenarios:
> 
> * If the new file path doesn't exist in the hash
>   
>   * Start consuming file.
>     
>     * **Or we can loop through the dictionary to check the fingerprints?? But this increases the time complexity and I don't think this is fruitful.**
> * If the new file path exists in the hash
>   
>   * Fetch the fingerprint for that path and check if it matches.
>   * If the fingerprint matches, then it means it's the same file as before and we can skip it.
>   * If the fingerprint doesn't match, it means it's a new file with the same name (the file is rotated) and we can start consuming it.
> 
> I'll implement it today and test it out Let me know what do think about this!

I'm not sure I see the whole picture here, but I get some of it. 

> New Hash: map[string]fingerprint

This is `map[filePath]fingerprint`, so basically "the last known fingerprint at this filepath". Is that right?

> * If the new file path doesn't exist in the hash
>   
>   * Start consuming file.

In your current implementation, you are deleting entries from the hash every time the eof is reached. So, when this condition is detected, all it tells us is that there _might_ be data to consume. However, we would still need to make that determination based on the fingerprint. (If we have an fingerprint that matches, use its offset.)


>   * If the fingerprint matches, then it means it's the same file as before and we can skip it.

I think it would mean that we should read to the end of the file, starting from the offset associated with the fingerprint.

>   * If the fingerprint doesn't match, it means it's a new file with the same name (the file is rotated) and we can start consuming it.

I agree it would indicate that the file we previously found here has been rotated. However, we do not know enough based on this information to determine how to handle the new file. It could be a file that we have already partially or fully consumed, so once again we need to check for a fingerprint match and if found consume from the offset.

Reply:
Yes, you're right.
We add the file path and fingerprint to the hash before starting to consume and remove it from the bash when we're done consuming.

Reply:
> > If the fingerprint doesn't match, it means it's a new file with the same name (the file is rotated) and we can start consuming it.
> 
> I agree it would indicate that the file we previously found here has been rotated. However, we do not know enough based on this information to determine how to handle the new file. It could be a file that we have already partially or fully consumed, so once again we need to check for a fingerprint match and if found consume from the offset.

Well, it means that the file is indeed new and we can start consuming it.
If it has already being consumed by previous poll cycles, then it will be added to `m.knownFiles` array and we will have it offseted as you mentioned.

Reply:
> Well, it means that the file is indeed new and we can start consuming it.

We may be talking about different things because I consider this to be incorrect.

I'm looking at this entire condition, "the new file path exists in the hash". My understanding is that what you meant by "the new file path" is that this is a file path that we've matched at the beginning of a poll interval, but we haven't yet determined whether or not we know about this file.

> * If the new file path exists in the hash
>   
>   * Fetch the fingerprint for that path and check if it matches.
>   * If the fingerprint matches, then it means it's the same file as before and we can skip it.
>   * If the fingerprint doesn't match, it means it's a new file with the same name (the file is rotated) and we can start consuming it.

When you say "the new file path exists in the hash", this only tells us that _a_ file existed in the same place and is currently being consumed. It does not account for cases where this file has been rotated _while we consume it_. A new file could be in its place _and_ we can only determine if we already have consumed this new file based on checking for known fingerprints.

Therefore, I believe this design would introduce unacceptable regressions.

Reply:
What I'm thinking is, adding a data structure, TRIE.
Which maps from fingerPrint->path.
This would cover the condition when a file has been rotated while we consume it.
We would first try to match the fingerprint in the TRIE:
if it didn't exist, it's straightforward that the file is new.
if it does, it means it's the old file and we're already consuming it. We can add another check for file path and proper logging instead.

Reply:
The reason to choose TRIE is that we can consider the fingerprint as a sequence of bytes and TRIE provides a great improvement over naive methods.

Reply:
I've considered storing fingerprints in a trie before and agree it is a natural structure because the fingerprints grow. However, it's not clear to me how we manage this structure over the longer term. That is, when/how do we remove old fingerprints so that we do not grow the trie indefinitely? Currently, we forget about fingerprints after we have not seen them for 3 poll cycles.

Reply:
If we use a fingerprint trie to keep track of the files we're consuming, what value is the file path?

The only value I can think of is that it could help us detect lost files. Basically, if the fingerprint already exists in the trie, but the path has changed, then we could hold onto the open file handle for longer. However, in this new design, we are already holding onto file handles for as long as possible anyways, so I'm not sure what the point is.

Reply:
How will the trie handle partial/growing fingerprints? 

Say we find a file with fingerprint `12345` and then in the next cycle we find `1234567890`. (Assume these are much much larger fingerprints, otherwise we would just finish consuming the file instantly.) 

If we only check for an exact match, we would conclude that this is a new file and we would end up consuming `12345` a second time. Are we able to use the trie in such a way that we can detect a partial match?

Reply:
> If we use a fingerprint trie to keep track of the files we're consuming, what value is the file path?

The only value I can think of is that it could help us detect lost files. Basically, if the fingerprint already exists in the trie, but the path has changed, then we could hold onto the open file handle for longer. However, in this new design, we are already holding onto file handles for as long as possible anyways, so I'm not sure why we care about paths at all.

Yeah, same thoughts.

Reply:
> How will the trie handle partial/growing fingerprints?
> 
> Say we find a file with fingerprint 12345 and then in the next cycle we find 1234567890. (Assume these are much much larger fingerprints, otherwise we would just finish consuming the file instantly.)
> 
> If we only check for an exact match, we would conclude that this is a new file and we would end up consuming 12345 a second time. Are we able to use the trie in such a way that we can detect a partial match?

I will use trie in such a way where we take care of this condtion.